Created Storage sample app with blob listing and MSAL

vcolin7 opened the pull request:
In collaboration with @anuchandy.

This depends on PR #174.

vcolin7 commented:
Marked as "Do Not Merge" as it depends on PR #174.

anuchandy commented:
@bsiegel - This is a sample app to list blobs in RecylerView control using PagedListAdapter. For now the paging abstractions that could be useful across client libs are in this sample (under sample\core dir). Once we agree on paging types to be in azure-core, we can move them. 

Sample is based on #174, so first that needs to be merged before we take this.

bsiegel commented on samples/sample-app-storage/build.gradle:
```diff
@@ -0,0 +1,56 @@
+    }
+}
+
+dependencies {
```
Let's get this section sorted (editor should be able to just do a "sort lines")

bsiegel commented on samples/sample-app-storage/src/main/java/com/azure/android/storage/sample/MsalClient.java:
```diff
@@ -0,0 +1,94 @@
+                    @Override
+                    public void onError(MsalException exception) {
+                        Log.d(TAG, "[acquireTokenSilentAsync]: Authentication failed: " + exception.toString());
+                        if (exception instanceof MsalClientException) {
```
Can we add a comment here about what this type of exception means?

anuchandy replied:
it's for general local error in msal library. 

@vcolin7 could you please add a comment referring to https://javadoc.io/doc/com.microsoft.identity.client/msal/0.1.1/com/microsoft/identity/client/MsalClientException.html

vcolin7 replied:
Sure thing

bsiegel commented on samples/sample-app-storage/src/main/java/com/azure/android/storage/sample/MsalClient.java:
```diff
@@ -0,0 +1,94 @@
+
+import java.util.Date;
+
+class MsalClient {
```
Nit: How do you feel about capitalizing this (and other MSAL-related classes) like `MSALClient`?

anuchandy replied:
Java guideline is to use pascal-case for acronyms so `Msal`.

anuchandy replied:
https://azure.github.io/azure-sdk/java_design.html#java-naming-uppercase-acronyms

bsiegel commented:
Seeing a lint issue that is most likely a bug:

```
android/samples/sample-app-storage/src/main/java/com/azure/android/storage/sample/ContainerBlobsPagedListAdapter.java:94: Error: Suspicious equality check: equals() is not implemented in com.azure.android.storage.blob.models.BlobItem [DiffUtilEquals]
              return oldItem.equals(newItem);
                             ~~~~~~~~~~~~~~~
```

anuchandy commented:
@vcolin7 for `oldItem.equals(newItem)` linting, we could just compare the {oldItem.name, newItem.name} and {oldItem.versionId, newItem.versionId} given blob names are unique within a container.

vcolin7 commented:
@bsiegel @anuchandy Added a link to the MsalException JavaDoc, fixed the lint error with Anu's suggestion and reordered the dependencies in our Gradle file. Also solved a small merge conflict.

anuchandy commented:
@vcolin7 thank you!, 2 minor feedbacks:

1. can we move the Exception doc reference to inside the `if`

```java
if (exception instanceof MsalClientException) {
    // HERE
}
```

2. please double check the last upstream sync to ensure Brandon's build script changes are there.

bsiegel commented:
Agreed with Anu about the exception feedback. I like the other comments that you have there because they tell me why that type of exception might be thrown. E.g. An `MsalServiceException` means "Exception when communicating with the STS, likely config issue".

I'd just like a similar comment that explains why an `MsalClientException` might occur. Something super brief, like "SDK was able to communicate with the STS, but was unable to authenticate successfully"

vcolin7 commented:
@bsiegel @anuchandy Re-applied the Gradle build fix and moved the MsalException comment to where suggested. I also added a bit more info on the other types of exception.

anuchandy commented on samples/sample-app-storage/src/main/java/com/azure/android/storage/sample/MsalClient.java:
```diff
@@ -40,15 +40,18 @@ public void onSuccess(IAuthenticationResult authResult) {
                         } else if (exception instanceof MsalUiRequiredException) {
-                            /* Tokens expired or no session, retry with interactive */
+                            // Exception thrown when auth token is expired or there is no active session, must retry
+                            // with interactive login. For more information, see: https://javadoc.io/static/com.microsoft.identity.client/msal/0.1.1/com/microsoft/identity/client/MsalServiceException.html
```
link should be : https://javadoc.io/static/com.microsoft.identity.client/msal/0.1.1/com/microsoft/identity/client/MsalUiRequiredException.html

bsiegel replied:
Good catch :)

vcolin7 replied:
You are right. I'll fix that real quick.
